url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/io/Generic_StaticIO.java

```diff
@@ -0,0 +1,2156 @@
+                e.printStackTrace(System.err);
+                System.exit(Generic_ErrorAndExceptionHandler.ClassNotFoundException);
+            }
+        }
+        return result;
+    }
+
+    /**
+     * @param input_File A File which is not a Directory to be copied
+     * @param outputDirectory_File The Directory to copy to.
+     */
+    private static void copyFile(
+            File input_File,
+            File outputDirectory_File) {
+        copyFile(
+                input_File,
+                outputDirectory_File,
+                input_File.getName());
+    }
+
+    /**
+     * @param fileToCopy A File which is not a Directory to be copied
+     * @param directoryToCopyInto A File that is the Directory to copy to.
+     * @param outputFileName The name for the file that will be created in the
+     * outputDirectory_File
+     */
+    public static void copyFile(
+            File fileToCopy,
+            File directoryToCopyInto,
+            String outputFileName) {
+        // String osName = System.getProperty( "os.name" );
+        // if ( osName.equalsIgnoreCase( "UNIX" ) ) {
+        // Runtime.getRuntime().exec( "cp " + _InputFile.toString() + " " +
+        // _Output_File.toString() );
+        // } else {
+        // }
+        if (!fileToCopy.exists()) {
+            System.err.println(
+                    "!input_File.exists() in "
+                    + Generic_StaticIO.class.getCanonicalName()
+                    + ".copy(File(" + fileToCopy + "),File(" + directoryToCopyInto + "))");
+            System.exit(Generic_ErrorAndExceptionHandler.IOException);
+        }
+        if (!directoryToCopyInto.exists()) {
+            directoryToCopyInto.mkdirs();
+        }
+        File output_File = new File(
+                directoryToCopyInto, outputFileName);
+        if (output_File.exists()) {
+            System.out.println(
+                    "Overwriting File " + output_File + " in "
+                    + Generic_StaticIO.class.getCanonicalName()
+                    + ".copy(File(" + fileToCopy + "),"
+                    + "File(" + directoryToCopyInto + "))");
+        } else {
+            try {
+                output_File.createNewFile();
+            } catch (IOException e) {
+                System.err.print(e.getMessage());
+                System.err.println(
+                        "Unable to createNewFile " + output_File + " in "
+                        + Generic_StaticIO.class.getCanonicalName()
+                        + ".copy(File(" + fileToCopy + "),"
+                        + "File(" + directoryToCopyInto + "))");
+                e.printStackTrace(System.err);
+                System.exit(Generic_ErrorAndExceptionHandler.IOException);
+            }
+        }
+        try {
+            BufferedInputStream bis;
+            bis = getBufferedInputStream(fileToCopy);
+            BufferedOutputStream bos;
+            bos = getBufferedOutputStream(output_File);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *bos*. The resource is referenced at lines 216, 219, 221. The resource is closed at line 222. There are other execution paths that do not contain closure statements, for example, when FilterInputStream.read throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).